In [4]:
# -*- coding: utf-8 -*-
"""
Created on Wed Jul 19 19:26:45 2017

@author: admin
"""

import pandas as pd
from difflib import SequenceMatcher
import numpy as np
#Function that computes the similarity ratio using Gestalt Pattern Matching algorithm
def similar(a,b):
    return SequenceMatcher(None,a,b).ratio()
#Read in the dataframes and define sub sequences as required   
df = pd.read_csv('C:/Users/admin/findata.csv')  
a = df.set_index('Name').index.get_loc(raw_input('Enter the Company name: ') )
training = df.iloc[:,2:11].fillna(value='0')
year = training[['Founding year']].astype(int).values
country = training['Company country']
training = training.drop(['Total funding','Number of employees','Founding year','Company country'],axis=1)
column_identifier = training.columns.tolist()
similarity_index =  np.zeros(shape = (df.shape[0] ,len(column_identifier)))
#compute the base similarity index using only Tags, Company Description, Product Description and partners features
for i in range(len(column_identifier)):
    for j in range(df.shape[0]):
        if training.loc[j,column_identifier[i]] == '0':
            similarity_index[j,i] = 0
        else :
            similarity_index[j,i] = similar(training.loc[a,column_identifier[i]],training.loc[j,column_identifier[i]])
[similarity_index[:,i]/max(similarity_index[:,i]) for i in range(len(column_identifier))]  
similarity_index = np.array([sum(similarity_index[i,:]) for i in range(df.shape[0])])
#compute the similarity when investor feature is checked
#def investor_filter():
investor_index = []
for j in range(df.shape[0]):
    if training.loc[j,'Investors'] == '0':
        investor_index.append(0)
    else:
        investor_index.append(similar(training.loc[a,'Investors'],training.loc[j,'Investors']))
investor_index[j] = investor_index[j]/max(investor_index)
investor_index = np.array([similarity_index[i]+investor_index[i] for i in range(df.shape[0])])
buff2 = investor_index.argsort()[-4:][::-1]
buff2 = np.delete(buff2,0)
print 'Filtering by investor:\n%s\n'%(df.loc[buff2,'Name'])
#compute the similarity when lifestage feature is checked
#def year_filter(company_name):
lifestage_index = []
for j in range(df.shape[0]):
    if abs(year[a]-year[j]) < 3:
        lifestage_index.append(1)
    else:
        lifestage_index.append(0)  
lifestage_index = np.array([similarity_index[i]+lifestage_index[i] for i in range(df.shape[0])]) 
buff3 = lifestage_index.argsort()[-4:][::-1]
buff3 = np.delete(buff3,0)
print 'Filtering by life stage:\n%s\n'%(df.loc[buff3,'Name'])
#compute the similarity when country feature is checked
country_index = []
for i in range(df.shape[0]):
    if similar(country.loc[a],country.loc[i]) <= 0.5:
        country_index.append(0)
    else:
        country_index.append(1)
country_index = np.array([similarity_index[i]+country_index[i] for i in range(df.shape[0])])
buff4 = country_index.argsort()[-4:][::-1]
buff4 = np.delete(buff4,0)
print 'Filtering by country:\n%s'%(df.loc[buff4,'Name'])




Enter the Company name: Capabiliti
Filtering by investor:
267         Trym
3569    Hivebeat
7651     CoinPip
Name: Name, dtype: object

Filtering by life stage:
307         ByeO
267         Trym
3569    Hivebeat
Name: Name, dtype: object

Filtering by country:
1814                GiftAssets
3803                 Connaizen
55      Datasigns Technologies
Name: Name, dtype: object


Enter the Company name: Bold
Filtering by investor:
4257         A-SaaS
7978    MineralTree
4204       Fusebill
Name: Name, dtype: object

Filtering by life stage:
4204    Fusebill
4482        Qwil
8137       DoPay
Name: Name, dtype: object

Filtering by country:
3843       Ceannate
7978    MineralTree
4482           Qwil
Name: Name, dtype: object

Enter the Company name: BTCjam
Filtering by investor:
6841         Purse
7764    Wayniloans
5558       Simplex
Name: Name, dtype: object

Filtering by life stage:
7764           Wayniloans
6841                Purse
3339    Crypto Facilities
Name: Name, dtype: object

Filtering by country:
6841         Purse
7564    Swarm Fund
7718      Snapcard
Name: Name, dtype: object

Enter the Company name: PPDai
Filtering by investor:
8280        BankFacil
4007         Duanrong
8018    Lending Works
Name: Name, dtype: object

Filtering by life stage:
6161        Yirendai
8285        AuxMoney
7209    Smartika Spa
Name: Name, dtype: object

Filtering by country:
2496    Yonglibao
6161     Yirendai
4007     Duanrong
Name: Name, dtype: object

Enter the Company name: Splitit
Filtering by investor:
7545                Paytopia
714                PiggyBank
7442    Circle Plus Payments
Name: Name, dtype: object

Filtering by life stage:
7413     PayDivvy
6361    Mobeewave
7704      Punchey
Name: Name, dtype: object

Filtering by country:
7545                Paytopia
7442    Circle Plus Payments
7413                PayDivvy
Name: Name, dtype: object

Enter the Company name: Qalize
Filtering by investor:
4189         LoopPay
6966    Buscaparking
6471           Zoona
Name: Name, dtype: object

Filtering by life stage:
4189           LoopPay
6966      Buscaparking
7797    Swish Payments
Name: Name, dtype: object

Filtering by country:
4189          LoopPay
7140           PAYTOO
8298    Advice Wallet
Name: Name, dtype: object

Enter the Company name: Alipay
Filtering by investor:
8037       Jusp
7576       Biyo
7310    instore
Name: Name, dtype: object

Filtering by life stage:
7262    Lightspeed POS
2910         Acclivity
7128     PivotalMOBILE
Name: Name, dtype: object

Filtering by country:
6487    QFPay
2491    Bypay
2460    Fuiou
Name: Name, dtype: object

Enter the Company name: Raspberry
Filtering by investor:
7497    HedgeChatter
2326       Paytailor
6642           Clipp
Name: Name, dtype: object

Filtering by life stage:
2326         Paytailor
6642             Clipp
7797    Swish Payments
Name: Name, dtype: object

Filtering by country:
2842         Vivonet
7497    HedgeChatter
7320     Hyperwallet
Name: Name, dtype: object

Enter the Company name: Coinding
Filtering by investor:
7651                 CoinPip
7955                 Monetsu
3181    Wishbird Experiences
Name: Name, dtype: object

Filtering by life stage:
7651          CoinPip
7806    Stellar Chain
7955          Monetsu
Name: Name, dtype: object

Filtering by country:
4868         Ripio
8098    enBitcoins
7764    Wayniloans
Name: Name, dtype: object

Enter the Company name: Giftly
Filtering by investor:
6952     CashU
8069     Flint
6522    Chillr
Name: Name, dtype: object

Filtering by life stage:
8069          Flint
7848    Prism Money
7871      Paynearme
Name: Name, dtype: object

Filtering by country:
8069          Flint
643           Verse
7848    Prism Money
Name: Name, dtype: object

Enter the Company name: Pandai
Filtering by investor:
5190         Tarya
8001         Loanz
7846    PylonLoans
Name: Name, dtype: object

Filtering by life stage:
8001         Loanz
7331      Harmoney
7063    SocietyOne
Name: Name, dtype: object

Filtering by country:
634     91Zhengxin
2496     Yonglibao
4007      Duanrong
Name: Name, dtype: object

Enter the Company name: Parametric
Filtering by investor:
6764              Pango
3382    Ignition Wealth
6722              VENDi
Name: Name, dtype: object

Filtering by life stage:
6778        Ooploo
7415        Unipay
6752    PayByGroup
Name: Name, dtype: object

Filtering by country:
6774          OpMoSys
6770           OrdrIt
8034    Knox Payments
Name: Name, dtype: object

Enter the Company name: Behalf
Filtering by investor:
5202          Prospa
8302    Able Lending
8186       Creditera
Name: Name, dtype: object

Filtering by life stage:
5202          Prospa
8186       Creditera
8302    Able Lending
Name: Name, dtype: object

Filtering by country:
8186       Creditera
8302    Able Lending
6396          Lendio
Name: Name, dtype: object

Enter the Company name: Elevate
Filtering by investor:
1676    GAN Integrity
1469          Dealflo
6921      Housing.com
Name: Name, dtype: object

Filtering by life stage:
1675        GoldenKey
1824    Loyalty Prime
1348            StatX
Name: Name, dtype: object

Filtering by country:
1676    GAN Integrity
1675        GoldenKey
1348            StatX
Name: Name, dtype: object

Enter the Company name: Bitwala
Filtering by investor:
5592                BITSO
3339    Crypto Facilities
8256                 BitX
Name: Name, dtype: object

Filtering by life stage:
5592                BITSO
3339    Crypto Facilities
8245                Magnr
Name: Name, dtype: object

Filtering by country:
6987    Bitmymoney
6604        Florin
8268      Bitex.la
Name: Name, dtype: object

Enter the Company name: Sinba
Filtering by investor:
6774         OpMoSys
6966    Buscaparking
7259         Lipisha
Name: Name, dtype: object

Filtering by life stage:
6966     Buscaparking
8069            Flint
7976    Minkasu, Inc.
Name: Name, dtype: object

Filtering by country:
4349    Aliena
8037      Jusp
1611    Pagita
Name: Name, dtype: object

Enter the Company name: Zoomdata
Filtering by investor:
1873    CROWDSURFER
1821        Seerene
4720      Aerospike
Name: Name, dtype: object

Filtering by life stage:
1873    CROWDSURFER
1821        Seerene
6094       SignalFx
Name: Name, dtype: object

Filtering by country:
3862    Syntelli Solutions Inc.
6094                   SignalFx
2427       Brand Big Data (BBD)
Name: Name, dtype: object

Enter the Company name: Toss
Filtering by investor:
7306    Inventure
6525       Masabi
6471        Zoona
Name: Name, dtype: object

Filtering by life stage:
7306      Inventure
6471          Zoona
7383    Activehours
Name: Name, dtype: object

Filtering by country:
3770    Spendwallet
5543      Coinhills
3760         N-Cash
Name: Name, dtype: object

Enter the Company name: CXA
Filtering by investor:
3966    ConnexionsAsia
8206     ComparaOnline
2009             Qover
Name: Name, dtype: object

Filtering by life stage:
1912    HeTai Life Insurance Co
2496                  Yonglibao
1524                 ThinkCloud
Name: Name, dtype: object

Filtering by country:
968             LiveSmart
1531    Digital Fineprint
3339    Crypto Facilities
Name: Name, dtype: object

Enter the Company name: Fundation
Filtering by investor:
5202    Prospa
6478    Behalf
6617     EzBob
Name: Name, dtype: object

Filtering by life stage:
5202        Prospa
6478        Behalf
8156    Dealstruck
Name: Name, dtype: object

Filtering by country:
6478         Behalf
8156     Dealstruck
7699    Bond Street
Name: Name, dtype: object